### PR TITLE
Clean up `rustc_features`

### PR DESCRIPTION
**Pull request by nnethercote**
Plenty more to be done, but this is a decent start.

r? @Nilstrieb

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -43,7 +37,15 @@ macro_rules! declare_features {
...
                 // (sym::$feature, $ver, $issue, $edition, set!($feature))
                 Feature {
-                    state: State::Active { set: set!($feature) },
+                    state: State::Active {
+                        // Sets this feature's corresponding bool within `features`.
+                        set: {
+                            fn f(features: &mut Features) {
+                                features.$feature = true;
+                            }
+                            f as fn(&mut Features)
```
Comment (petrochenkov) [inline compiler/rustc_feature/src/active.rs:46]: Capture-less closures coerce into fn pointers, so you probably don't need a separate function or an explicit cast here.

Comment (nnethercote) [reply compiler/rustc_feature/src/active.rs:46]: You are right that the cast can be removed. But I can't see how to remove the separate named function.

Comment (nnethercote) [reply compiler/rustc_feature/src/active.rs:46]: Oh, use a closure, I see. Good suggestion.

Comment (nnethercote) [reply compiler/rustc_feature/src/active.rs]: I added a new commit converting this to a closure. Thanks!

**Review by Noratrieb — APPROVED** [review]

**Pull request merged**